### PR TITLE
PSK key exchange

### LINKED ISSUE #265: Pure PSK key exchange in TLS 1.3
add support for `psk_ke` mode in TLS 1.3 handshake, both for client support and server support, continue defaulting to `psk_dhe_ke`

related extension: `pre_shared_key`

### PR DESCRIPTION
**Pull request by tomato42**
fixes #265

---
This change is [Reviewable](https://reviewable.io/reviews/tomato42/tlslite-ng/266)

### CODE REVIEW

**Comment by tomato42** [comment]
This pull request **introduces 1 alert** and **fixes 1** when merging 34bf78db51249a3ef33c60f8d8a1f7d1b4b31123 into 5f6110b7ca6a19fdd79bc64f230526df30ee6c0f - [view on lgtm.com](https://lgtm.com/projects/g/tomato42/tlslite-ng/rev/pr-3b6e6de495ca300ffcb888301f186bd1c8e9f2ec)

**new alerts:**

* 1 for \_\_eq\_\_ not overridden when adding attributes

**fixed alerts:**

* 1 for \_\_eq\_\_ not overridden when adding attributes

---

*Comment posted by [lgtm.com](https://lgtm.com)*

File: unit_tests/test_tlslite_handshakesettings.py [hunk 1]
```diff
@@ -369,6 +369,15 @@ def test_ticketLifetime_wrong(self):
...
+        hs.psk_modes = ["psk_pqe_ke"]
+
+        with self.assertRaises(ValueError) as e:
+            hs.validate()
+
+        self.assertIn("psk_pqe_ke", str(e.exception))
+
 
 if __name__ == '__main__':
     unittest.main()
```
Comment (nmav) [inline unit_tests/test_tlslite_handshakesettings.py]: r+

File: tlslite/tlsconnection.py [hunk 2]
```diff
@@ -3414,15 +3442,14 @@ def _handshakeWrapperAsync(self, handshaker, checker):
...
-                raise TLSMissingExtension("Signature algorithms extension"
-                                          "missing")
-            if not hashAndAlgsExt.sigalgs:
-                raise TLSDecodeError("Signature algorithms extension empty")
+                # the error checking was done before hand, likely we're
+                # doing PSK key exchange
+                return
 
         if hashAndAlgsExt is None or hashAndAlgsExt.sigalgs is None:
             # RFC 5246 states that if there are no hashes advertised,
```
Comment (nmav) [inline tlslite/tlsconnection.py]: r+

**Comment by tomato42** [comment]
@nmav Thanks for the review

**Pull request merged**